packages/flutter_tools/lib/src/android/adb.dart

```diff
@@ -112,6 +118,26 @@ class Adb {
     return controller.stream;
   }
 
+  Future _populateDeviceNames(List<AdbDevice> devices) async {
+    for (AdbDevice device in devices) {
+      if (device.modelID == null) {
+        // If we don't have a name of a device in our cache, call `device -l` to populate it.
+        if (_idToNameCache[device.id] == null)
+          await _populateDeviceCache();
+
+        // Set the device name from the cached name. Adb device notifications only
+        // have IDs, not names. We get the name by calling `listDevices()`.
+        device.modelID = _idToNameCache[device.id];
+      }
+    }
+  }
+
+  Future _populateDeviceCache() async {
+    List<AdbDevice> devices = await listDevices();
+    for (AdbDevice device in devices)
+      _idToNameCache[device.id] = device.modelID;
+  }
+
```

Comment:
This code depends on us not clearing the cache between requests, and on device names not changing. I don't think there's a race condition here. Every request that needs names for A and B will wait until they've been populated in the cache before using them. Future (or parallel) requests will only add to the cache.

Happy to clarify or re-write this code if people see a better way :)